src/channel.cr

```diff
@@ -373,8 +425,8 @@ class Channel(T)
       return ops.size, NotReady.new
     end
 
-    state = Atomic(SelectState).new(SelectState::Active)
-    contexts = ops.map &.create_context_and_wait(pointerof(state))
+    state_ptr = Pointer(Atomic(SelectState)).malloc(1, Atomic(SelectState).new(SelectState::Active))
```

Comment:
Why we need for the state to be in the heap instead of the stack? The fiber will continue from this point and that was the reason we can use the stack.

Reply:
Because I move the cleaning operations out of lock section during close. Consider the case. A `channel` is closing and needs to clean up the **long list**. And a `select` operation has been done at the same time. The `select` had a `Sender` or `Receiver` in the **list**. The action will not be removed because we don't know the action has been iterated during close. If it iterate the action after the `select` return, the `select_context.try_trigger` may modify the invaild address. Hence, the simple solution is put the state in the heap.

Reply:
Great. Can we add that as a comment at this point in the code?

Reply:
Sure, I will add some description about it.